brcm63xx: add support for ADB P.DG A4101N A-000-1A1-AE

DanCasterIt opened the pull request:
ADB P.DG A4101N A-000-1A1-AE has a similar PCB as the OpenWrt's ADB P.DG A4001N1
with LEDs connected to different GPIO PINs in active low configuration.

Hardware:
* Board ID: 96328avngv
* SoC: Broadcom BCM6328 (rev b0) @ 320MHz, CPU BMIPS4350
* RAM DDR2: 64 Mbyte - Winbond W9751G6KB-25
* Serial flash: 16 Mbyte - Winbond 25Q128BVFG
* Ethernet: 4x Ethernet 10/100 baseT
* Wifi 2.4GHz: Broadcom Corporation BCM43225 Wireless Network Adapter (rev 23)
* LEDs: 1x Power, 1x DSL, 1x Internet, 4x ETH, 1x USB, 1x WLAN, 1x WPS, 1x TEL
* Buttons: 1x Reset, 1x WPS, 1x unnamed
* UART: 1x TTL 115200n8, VCC RX TX GND, on J502 connector

Installation via CFE:
* Stock CFE has to be overwritten with a generic 6328 one that can upload
  .bin images with no signature check (cfe6328_configured.bin)
* Connect a serial port to the board
* Stop the cfe boot process after power on by pressing enter
* Set static IP 192.168.2.10 and subnet mask 255.255.255.0
* Navigate to http://192.168.2.50/
* Upload the OpenWrt image file

A4101N GPIO LAYOUT:

Power            always on
DSL              GPIO483(03)
Internet         GPIO491(11)
ETH1             GPIO505(25)
ETH2             GPIO506(26)
ETH3             GPIO507(27)
ETH4             GPIO508(28)
USB              GPIO490(10)
WLAN             controlled by BCM43225
WPS              GPIO489(09)
TEL              GPIO511(31)

Key RESET        GPIO503(23)
Key WPS          GPIO504(24)
Key unnamed      GPIO492(12)

Signed-off-by: Daniele Castro <danielecastro@hotmail.it>

DanCasterIt committed:
bcm63xx: add support for ADB P.DG A4101N A-000-1A1-AE

ADB P.DG A4101N A-000-1A1-AE has a similar PCB as the OpenWrt's ADB P.DG A4001N1
with LEDs connected to different GPIO PINs in active low configuration.

Hardware:
* Board ID: 96328avngv
* SoC: Broadcom BCM6328 (rev b0) @ 320MHz, CPU BMIPS4350
* RAM DDR2: 64 Mbyte - Winbond W9751G6KB-25
* Serial flash: 16 Mbyte - Winbond 25Q128BVFG
* Ethernet: 4x Ethernet 10/100 baseT
* Wifi 2.4GHz: Broadcom Corporation BCM43225 Wireless Network Adapter (rev 23)
* LEDs: 1x Power, 1x DSL, 1x Internet, 4x ETH, 1x USB, 1x WLAN, 1x WPS, 1x TEL
* Buttons: 1x Reset, 1x WPS, 1x unnamed
* UART: 1x TTL 115200n8, VCC RX TX GND, on J502 connector

Installation via CFE:
* Stock CFE has to be overwritten with a generic 6328 one that can upload
  .bin images with no signature check (cfe6328_configured.bin)
* Connect a serial port to the board
* Stop the cfe boot process after power on by pressing enter
* Set static IP 192.168.2.10 and subnet mask 255.255.255.0
* Navigate to http://192.168.2.50/
* Upload the OpenWrt image file

A4101N GPIO LAYOUT:

Power            always on
DSL              GPIO483(03)
Internet         GPIO491(11)
ETH1             GPIO505(25)
ETH2             GPIO506(26)
ETH3             GPIO507(27)
ETH4             GPIO508(28)
USB              GPIO490(10)
WLAN             controlled by BCM43225
WPS              GPIO489(09)
TEL              GPIO511(31)

Key RESET        GPIO503(23)
Key WPS          GPIO504(24)
Key unnamed      GPIO492(12)

Signed-off-by: Daniele Castro <danielecastro@hotmail.it>
[Amend commit description, DTS improvements, refresh patches]
Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented on target/linux/bcm63xx/base-files/etc/board.d/01_leds:
```diff
@@ -31,6 +31,11 @@ adb,pdg-a4001n-a-000-1a1-ax)
 	;;
+adb,pdg-a4101n-a-000-1a1-ae)
+	ucidef_set_led_netdev "lan" "LAN" "green:internet" "eth0.1"
+	ucidef_set_led_netdev "wan" "WAN" "green:dsl" "eth0.2"
```
There's no eth0.2 in the configuration since all (4) ethernet ports are added to eth0.1.

Noltari commented on target/linux/bcm63xx/dts/bcm6328-adb-pdg-a4101n-a-000-1a1-ae.dts:
```diff
@@ -0,0 +1,140 @@
+		poll-interval = <20>;
+
+		unnamed {
+			label = "unnamed_button";
```
What's the stock firmware functionality of this button?
Adding it to OpenWrt as "unnamed_button" doesn't feel right...

DanCasterIt replied:
Unfortunately when I got this router the stock firmware was corrupted and the router didn't boot. Anyway this button is the only one that has no label. In the official user manual this button is not mentioned at all.

I know it doesn't feel right but it's a an unnamed button for real...

Noltari commented on target/linux/bcm63xx/dts/bcm6328-adb-pdg-a4101n-a-000-1a1-ae.dts:
```diff
@@ -0,0 +1,140 @@
+	led@2 {
+		reg = <2>;
+		active-low;
+		label = "internet_led_enable";
```
If this enables and disables the internet LED, maybe we should add it as a GPIO enabled by default, not as a LED.
You should also add a gpio-leds group with value 31 for the TEL LED since GPIO31 can't be controlled with the LED driver. See https://github.com/openwrt/openwrt/blob/master/target/linux/bcm63xx/dts/bcm6362-netgear-dgnd3700-v2.dts#L52

Noltari commented on target/linux/bcm63xx/dts/bcm6328-adb-pdg-a4101n-a-000-1a1-ae.dts:
```diff
@@ -0,0 +1,140 @@
+				label = "cfe";
+				read-only;
+
+				partition@580 {
```
Please, remove this partition.

Noltari commented on target/linux/bcm63xx/patches-5.4/805-board_bcm6328-adb-pdg-a4101n-a-000-1a1-ae.patch:
```diff
@@ -0,0 +1,70 @@
+Index: linux-5.4.63/arch/mips/bcm63xx/boards/board_bcm963xx.c
+===================================================================
+--- linux-5.4.63.orig/arch/mips/bcm63xx/boards/board_bcm963xx.c
++++ linux-5.4.63/arch/mips/bcm63xx/boards/board_bcm963xx.c
```
Please, refresh the patch (make target/linux/refresh).
Also, this should now be integrated into https://github.com/openwrt/openwrt/blob/master/target/linux/bcm63xx/patches-5.4/512-board_bcm6328.patch

Noltari commented on target/linux/bcm63xx/base-files/etc/board.d/02_network:
```diff
@@ -21,7 +21,11 @@ t-com,speedport-w-500v)
 adb,a4001n|\
 adb,pdg-a4001n-a-000-1a1-ax|\
-adb,pdg-a4101n-a-000-1a1-ae|\
+adb,pdg-a4101n-a-000-1a1-ae)
```
Here you are also changing the network configuration of adb,a4001n1, adb,a4001n, adb,pdg-a4001n-a-000-1a1-ax... and we don't want to do that.

DanCasterIt replied:
I fixed the problem. Sorry

Noltari commented:
@DanCasterIt can you test my branch?
- https://github.com/Noltari/openwrt/commits/bcm63xx-adb-pdg-a4101n-a-000-1a1-ae
- https://github.com/Noltari/openwrt/commit/0b54f9d16d91014586a71658676938af45d0ea11

I've merged and refreshed the board patches and performed some modifications in order to accept these patches:
- Removed LAN/WAN LEDs, since the device has its own ethernet LEDs and I suppose that none of them is labelled as WAN (all of them are intended for LAN with stock firmware). Therefore, all ethernet ports have been assigned to eth0.1.
- Use gpio-hog for gpio 2 (green internet led switch).
- Merge board patches into bcm6328 board patch.
- Rename _unnamed_ button to wifi.

adschm commented:
@Noltari Consider changing the gpio-hog label from "green:internet:switch" to "green:internet-switch" or similar, since the three colon-separated values imply model:color:function, which is not met here.

Noltari commented:
> @Noltari Consider changing the gpio-hog label from "green:internet:switch" to "green:internet-switch" or similar, since the three colon-separated values imply model:color:function, which is not met here.

Done 👍

DanCasterIt commented:
@Noltari I compiled an image from your branch and it works with the only problem that the DSL LED stuks at on. I fixed the problem changing "active-high" in the dts file.

I also re-added and corrected the WAN stuff (LEDs and network files). Please feel free to remove these modifications again if you think they won't be good.
I suggest to add the those modifications because in the stock firmware the "DSL" and "Internet" LEDs are related to DSL WAN activity.

In the ADB P.DG A4001N A-000-1A1-AX pull request (https://github.com/openwrt/openwrt/pull/2598) you merged few months ago there are WAN LEDs with no WAN configuration like this pull request was

Noltari commented:
> @Noltari I compiled an image from your branch and it works with the only problem that the DSL LED stuks at on despite being specified "default-state = "off"" in the dts.
> I checked, it's in active-low configuration, so I don't know what's wrong.

This should be solved by removing _led-running_ from the DT aliases.

> I also re-added and corrected the WAN stuff (LEDs and network files). Please feel free to remove these modifications again if you think they won't be good.
> I suggest to add those modifications because in the stock firmware the "DSL" and "Internet" LEDs are related to DSL WAN activity.

I would like to know if any of the ethernet ports is labelled as "wan", "fibre", "internet" or something that indicates non-LAN behavior. If this is not the case, it's better to replicate the stock firmware and add all ethernet ports as LAN ports. After that, the user can change that default behavior and change one of the ports to act as WAN.
The issue here is that OpenWrt doesn't support xDSL for bcm63xx devices, so we can't completely replicate stock.

> In the ADB P.DG A4001N A-000-1A1-AX pull request (#2598) you merged few months ago there are WAN LEDs with no WAN configuration like this pull request was

Yeah, I know. I had already realized of that and was going to fix it after merging this PR.

DanCasterIt commented:
>> @Noltari I compiled an image from your branch and it works with the only problem that the DSL LED stuks at on despite being specified "default-state = "off"" in the dts.
>> I checked, it's in active-low configuration, so I don't know what's wrong.

> This should be solved by removing _led-running_ from the DT aliases.

You're right. I solved the problem.

> I would like to know if any of the ethernet ports is labelled as "wan", "fibre", "internet" or something that indicates non-LAN behavior.

The ethernet ports are all labelled as ETH1-4.
> If this is not the case, it's better to replicate the stock firmware and add all ethernet ports as LAN ports. After that, the user can change that default behavior and change one of the ports to act as WAN.
> The issue here is that OpenWrt doesn't support xDSL for bcm63xx devices, so we can't completely replicate stock.

Exactly, that's why I thought to add a WAN port: because the router is supposed to have at least one WAN port (the xDSL) and adding one to the switch would better (in my opinion) replicate the stock FW behaviour.
Without adding a WAN to the switch, the user would have a stock OpenWrt router with no WANs at all that, I think, does not make much sense, it would be not ready to use for sure. Further when I was an OpenWrt noob I took quite few time to figure out about VLANs and how to add a WAN ports. In this way I think that the image would be a ready to use one to simplify the user's experience.

> > In the ADB P.DG A4001N A-000-1A1-AX pull request (#2598) you merged few months ago there are WAN LEDs with no WAN configuration like this pull request was
> 
> Yeah, I know. I had already realized of that and was going to fix it after merging this PR.

At this point I would suggest also to wait to merge my other two PRs for which I'm waiting to merge this to be able to correctly refresh the patches:
https://github.com/openwrt/openwrt/pull/3393
https://github.com/openwrt/openwrt/pull/3400

Noltari commented:
Merged, thanks!
https://github.com/openwrt/openwrt/commit/61ef5940f2eafa2a883d35af3bbf4622722c2fbc
